lib/ProgramCall.js

```diff
@@ -31,59 +33,61 @@ class ProgramCall {
+   * Internal function to handle ds and data within <parm> node
+   * There is an open propsal to add private methods to JS.
+   * https://github.com/tc39/proposal-private-methods
+   * We should update to use private methods in the future.
+   * @param {object} data
+   */
+  __addData__(data = {}) {
+    if (data.type === 'ds') {
       this.xml += '<ds';
-      if (opt && typeof opt === 'object') { // append <ds> options
-        if (opt.dim) { this.xml += ` dim='${opt.dim}'`; }
-        if (opt.dou) { this.xml += ` dou='${opt.dou}'`; }
-        if (opt.len) { this.xml += ` len='${opt.len}'`; }
-        if (opt.data) { this.xml += ` data='${opt.data}'`; }
-      }
+      // append <ds> options
+      if (data.dim) { this.xml += ` dim='${data.dim}'`; }
+      if (data.dou) { this.xml += ` dou='${data.dou}'`; }
+      if (data.len) { this.xml += ` len='${data.len}'`; }
+      if (data.data) { this.xml += ` data='${data.data}'`; }
       this.xml += '>';
 
-      for (let i = 0; i < data.length; i += 1) {
-        this.addParam(data[i][0], data[i][1], data[i][2], true);
+      for (let i = 0; i < data.fields.length; i += 1) {
+        this.__addData__(data.fields[i]);
       }
       this.xml += '</ds>';
-    } else { // A simple parameter
-      this.xml += `<data type='${type}'`;
-      if (opt && typeof opt === 'object') { // append <data> options
-        Object.keys(opt).forEach((key) => {
-          this.xml += ` ${key}='${opt[key]}'`;
-        });
-      }
-      this.xml += `>${data}</data>`;
+    } else {
+      this.xml += `<data type='${data.type}'`;
+      // append <data> options
+      if (data.name) { this.xml += ` name='${data.name}'`; }
+      if (data.varying) { this.xml += ` varying='${data.varying}'`; }
+      if (data.enddo) { this.xml += ` enddo='${data.enddo}'`; }
+      if (data.setlen) { this.xml += ` setlen='${data.setlen}'`; }
+      if (data.offset) { this.xml += ` offset='${data.offset}'`; }
+      if (data.hex) { this.xml += ` hex='${data.hex}'`; }
+      if (data.trim) { this.xml += ` trim='${data.trim}'`; }
+      if (data.next) { this.xml += ` next='${data.next}'`; }
+      this.xml += `>${data.value}</data>`;
     }
+  }
 
-    if (!inDs) { // In recursive mode, if it is an element in DS, then no <parm> or </parm> needed.
-      this.xml += '</parm>';
+  /**
+  * @description adds a parameter to the program XML
+  * @param {object} parmeter
+  */
+  addParam(parameter = {}) {
+    if (typeof parameter !== 'object') {
+      throw new Error('Expected the first parameter to be an object');
+    }
+    if (parameter.type === undefined) {
+      throw new Error('Expected \'type\' key on the parameter object');
     }
+    this.xml += '<parm';
+
+    if (parameter.name) { this.xml += ` name=${parameter.name}`; }
+    if (parameter.io) { this.xml += ` io='${parameter.io}'`; }
+    if (parameter.by) { this.xml += ` by='${parameter.by}'`; }
+
+    this.xml += '>';
+    this.__addData__(parameter);
```

Comment:
Just thought of an issue. We're passing the attributes of the parameter and the top-level data item in the same object. This means that the name attribute is the same for the parameter _and_ the data item (currently we don't add the name for ds fields, though we probably should). How does this affect the generated JSON that we get back?

Should we remove the name attribute before calling `__addData__` so that the top-level data/ds does for the parameter is unnamed? Or does it make more sense to ensure that the parameter and data/ds have the same name.

Reply:
You brought up an interesting point.

First off we should allow ds to have a name. 

We should allow data and ds names to be configured independently of parameter name.

To do this we can add some additional keys. `dataName` and `dsName`.

What do you think?

Reply:
Why would the data and ds have a different name than the parameter? What benefit / use case would that solve?

Reply:
hmm I do not have a scenario where parameter name and ds / data would differ.

In that case, there is probably not much benefit ensuring the parameter and data/ds have the same name.

Both will be a child of the `<parm>` node anyway.

If a specific use case comes up we can always add additionally keys.

So we should remove the name attribute before calling `__addData__`.

Reply:
What does xml2json give you back in each situation?

Reply:
output [without removing the name attribute](https://gist.github.com/abmusse/019352e5a9fe4ab4b54b3de388bbec47) before calling `__addData__`.

output [with removing the name attribute](https://gist.github.com/abmusse/a85dcf331a2de705be0f6abbdf834c99) before calling `__addData__`.

Reply:
> What does xml2json give you back in each situation?

Did you mean xml2js or xmlToJson deprecated function?

Reply:
Sorry, yes xml2js

Reply:
So basically you get

```
{
    "$": {
        "name": "number_sys_values_to_retrieve"
    },
    "data": [
        {
        "_": "1",
        "$": {
            "type": "10i0",
            "name": "number_sys_values_to_retrieve"
        }
        }
    ]
},
```

or
```

{
    "$": {
        "name": "number_sys_values_to_retrieve"
    },
    "data": [
        {
        "_": "1",
        "$": {
            "type": "10i0",
        }
        }
    ]
},
```

Since there's that stupid data/ds node, I think it makes sense to have the name duplicated. Please add the name to the ds as well.

Reply:
OK will add the name to ds as well.

Reply:
updated [output](https://gist.github.com/abmusse/ba42ea78620af080b438a5c2ad5ca76e) with ds names